crypto/src/merkle/proofs.rs

```diff
@@ -255,6 +255,168 @@ impl<H: Hasher> BatchMerkleProof<H> {
+                        return Err(MerkleTreeError::InvalidProof);
+                    }
+                    buf[0] = self.nodes[i][0];
+                    match index_map.get(&(index + 1)) {
+                        Some(&index2) => {
+                            if self.leaves.len() <= index2 {
+                                return Err(MerkleTreeError::InvalidProof);
+                            }
+                            buf[1] = self.leaves[index2];
+                        }
+                        None => return Err(MerkleTreeError::InvalidProof),
+                    }
+                    proof_pointers.push(1);
+                }
+            }
+
+            // hash sibling nodes into their parent and add it to partial_tree
+            let parent = H::merge(&buf);
+            partial_tree_map.insert(offset + index, buf[0]);
+            partial_tree_map.insert((offset + index)^ 1, buf[1]);
+            let parent_index = (offset + index) >> 1;
+            v.insert(parent_index, parent);
+            next_indexes.push(parent_index);
+            partial_tree_map.insert(parent_index, parent);
+        }
+
+        // iteratively move up, until we get to the root
+        for _ in 1..self.depth {
+            let indexes = next_indexes.clone();
+            next_indexes.truncate(0);
+
+            let mut i = 0;
+            while i < indexes.len() {
+                let node_index = indexes[i];
+                let sibling_index = node_index ^ 1;
+
+                // determine the sibling
+                let sibling: H::Digest;
+                if i + 1 < indexes.len() && indexes[i + 1] == sibling_index {
+                    sibling = match v.get(&sibling_index) {
+                        Some(sibling) => *sibling,
+                        None => return Err(MerkleTreeError::InvalidProof),
+                    };
+                    i += 1;
+                } else {
+                    let pointer = proof_pointers[i];
+                    if self.nodes[i].len() <= pointer {
+                        return Err(MerkleTreeError::InvalidProof);
+                    }
+                    sibling = self.nodes[i][pointer];
+                    proof_pointers[i] += 1;
+                }
+
+                // get the node from the map of hashed nodes
+                let node = match v.get(&node_index) {
+                    Some(node) => node,
+                    None => return Err(MerkleTreeError::InvalidProof),
+                };
+
+                // compute parent node from node and sibling
+                if node_index & 1 != 0 {
+                    buf[0] = sibling;
+                    buf[1] = *node;
+                    partial_tree_map.insert(node_index ^ 1, sibling);
+                } else {
+                    buf[0] = *node;
+                    buf[1] = sibling;
+                    partial_tree_map.insert(node_index ^ 1, sibling);
+                }
+                let parent = H::merge(&buf);
```

Comment:
Both branches seem to put the same thing into the partial tree map (`partial_tree_map.insert(node_index ^ 1, sibling)`). If this is correct, you might be able to simplify this as:
```Rust
partial_tree_map.insert(node_index ^ 1, sibling);
let parent = if node_index & 1 != 0 {
    H::merge(&[sibling, *node])
} else {
    H::merge(&[*node, sibling])
};
```

Reply:
Indeed!